### PR TITLE
Look at target before moving

### PR DESCRIPTION
**Pull request by guiccbr**
Send the MAV_CMD_CONDITION_YAW command in order to have the vehicle continually
pointing at the direction of movement.

### CODE REVIEW

File: gzsitl/gzsitl_plugin.cc [hunk 1]
```diff
@@ -22,6 +22,8 @@
 #define GZSITL_PERM_TARG_POSE_PUB_FREQ_HZ 50
 #define GZSITL_VEHICLE_POSE_PUB_FREQ_HZ 50
 #define GZSITL_SUBS_TARG_POSE_SUB_MAX_RESPONSE_TIME 1000
+#define GZSITL_LOOKAT_TARG_ANG_LIMIT 30.0
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: name ir GZSITL_LOOKAT_TARGET_ANG_LIMIT. These can be static const values

File: gzsitl/gzsitl_plugin.cc [hunk 2]
```diff
@@ -171,6 +173,22 @@ void GZSitlPlugin::OnUpdate()
             subs_target->SetWorldPose(tpose_new);
         }
 
+        // Check if current target position is on the field of view. If not,
+        // stop in the current position and rotate towards target before moving
+        // forward.
+        math::Pose rel_pose = tpose_new - mpose_new;
+
+        double targ_ang = atan2(rel_pose.pos.y, rel_pose.pos.x);
+        targ_ang *= 180.0 / M_PI;
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: Merge this with the row above. Or use a rad2deg function

Comment (guiccbr) [inline gzsitl/gzsitl_plugin.cc]: Thanks. I've got those lines merged on commit 32a6703eb0deeb17a98b89309879458870fd08fe.

File: gzsitl/gzsitl_plugin.cc [hunk 3]
```diff
@@ -47,6 +47,56 @@ void GZSitlPlugin::OnUpdate()
...
+
+    // Publish current gazebo vehicle pose
+    math::Pose vehicle_pose = model->GetWorldPose();
+    this->vehicle_pose_pub->Publish(msgs::Convert(vehicle_pose.Ign()));
+
+    // Get pointer to the permanent target if exists
+    this->perm_target = model->GetWorld()->GetModel(perm_target_name);
+    if(this->perm_target) {
+        this->target_exists = true;
+    }else {
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: Missing space

File: gzsitl/gzsitl_plugin.cc [hunk 4]
```diff
@@ -47,6 +47,56 @@ void GZSitlPlugin::OnUpdate()
...
+        model->SetWorldPose(curr_pose);
+    }
+
+    // Publish current gazebo vehicle pose
+    math::Pose vehicle_pose = model->GetWorldPose();
+    this->vehicle_pose_pub->Publish(msgs::Convert(vehicle_pose.Ign()));
+
+    // Get pointer to the permanent target if exists
+    this->perm_target = model->GetWorld()->GetModel(perm_target_name);
+    if(this->perm_target) {
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: Missing space

File: gzsitl/gzsitl_plugin.cc [hunk 5]
```diff
@@ -47,6 +47,56 @@ void GZSitlPlugin::OnUpdate()
...
+
+    // Check if target is being overridden by the substitute target
+    if (is_target_overridden()) {
+        target_pose = get_subs_target_pose();
+        this->target_exists = true;
+    }
+
+    // Set substitute target position in Gazebo if exists
+    this->subs_target = model->GetWorld()->GetModel(subs_target_name);
+    if (subs_target) {
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: add this

File: gzsitl/gzsitl_plugin.cc [hunk 6]
```diff
@@ -348,11 +387,12 @@ void GZSitlPlugin::set_global_pos_coord_system(
...
 }
 
-void GZSitlPlugin::calculate_pose(math::Pose *pose, mavlink_attitude_t attitude,
-                                  mavlink_local_position_ned_t local_position)
+math::Pose GZSitlPlugin::calculate_pose(mavlink_attitude_t attitude,
+                             mavlink_local_position_ned_t local_position)
 {
-    pose->Set(local_position.x, -local_position.y, -local_position.z,
-              attitude.roll, -attitude.pitch, -attitude.yaw);
+    math::Pose tpose(local_position.x, -local_position.y, -local_position.z,
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: Add comments to explain why the -

Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: return directly instead of using tpose

File: gzsitl/gzsitl_plugin.cc [hunk 7]
```diff
@@ -47,6 +47,56 @@ void GZSitlPlugin::OnUpdate()
...
+
+    // Set substitute target position in Gazebo if exists
+    this->subs_target = model->GetWorld()->GetModel(subs_target_name);
+    if (subs_target) {
+        subs_target->SetWorldPose(target_pose);
+    }
+
+    // Calculate target relative pose if exists
+    math::Pose target_pose_rel;
+    if(this->target_exists) {
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: ditto

File: gzsitl/gzsitl_plugin.cc [hunk 8]
```diff
@@ -130,72 +176,45 @@ void GZSitlPlugin::OnUpdate()
...
-                0, 2000);
-            tpose = mpose_new;
-        }
+        // Store static target pose to avoid unnecessary repetition of requests
+        static math::Pose target_pose_prev = math::Pose::Zero;
 
         // Send Target if exists and if it has been moved
-        if (is_flying() && tpose_new != tpose) {
-            tpose = tpose_new;
+        if (is_flying() && target_pose != target_pose_prev) {
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc:221]: Why is_flying makes difference?

File: gzsitl/gzsitl_plugin.cc [hunk 9]
```diff
@@ -219,6 +238,26 @@ void GZSitlPlugin::OnUpdate()
...
+        if (fabs(targ_ang) <= GZSITL_LOOKAT_TARG_ANG_LIMIT) {
+            simstate = ACTIVE_AIRBORNE;
+            print_debug_state("state: ACTIVE_AIRBORNE\n");
+        }
+
+        // Otherwise, continue to request the rotation
+        mavserver.queue_send_cmd_long_until_ack(
+            MAV_CMD_CONDITION_YAW, fabs(targ_ang),
+            GZSITL_LOOKAT_ROT_SPEED_DEGPS, -copysign(1, targ_ang), 1, 0, 0, 0,
+            2000);
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: What is 2000?

File: gzsitl/gzsitl_plugin.cc [hunk 10]
```diff
@@ -219,6 +238,26 @@ void GZSitlPlugin::OnUpdate()
         return;
     }
 
+    case ACTIVE_ROTATING: {
+
+        // Calculate the target azimuthal angle of the target in relation to
+        // the vehicle
+        double targ_ang = (180.0 / M_PI) *
```
Comment (dakerfp) [inline gzsitl/gzsitl_plugin.cc]: create a rad2deg function

File: gzsitl/gzsitl_plugin.cc (hunk 1 above)
Comment (guiccbr) [inline gzsitl/gzsitl_plugin.cc]: I've put them as const values inside a local namespace named 'defaults'.

File: gzsitl/gzsitl_plugin.cc (hunk 3 above)
Comment (guiccbr) [inline gzsitl/gzsitl_plugin.cc]: good catch :)

File: gzsitl/gzsitl_plugin.cc (hunk 6 above)
Comment (guiccbr) [inline gzsitl/gzsitl_plugin.cc]: Comment added. This is to convert from NED to ENU (https://en.wikipedia.org/wiki/North_east_down)

File: gzsitl/gzsitl_plugin.cc (hunk 8 above)
Comment (guiccbr) [inline gzsitl/gzsitl_plugin.cc:221]: This allows a waypoint to be sent if and only if the vehicle is flying. When the vehicle is on ground, sending the target position is not necessary.

File: gzsitl/gzsitl_plugin.cc (hunk 9 above)
Comment (guiccbr) [inline gzsitl/gzsitl_plugin.cc]: It's a "retry interval", in ms. These calls to queue_send_cmd_long_until_ack shall all be encapsulated inside the new mavserver (mav_vehicles) on my next pr, but I'll fix it anyway adding a #define for this constant.

File: gzsitl/gzsitl_plugin.cc (hunk 10 above)
Comment (guiccbr) [inline gzsitl/gzsitl_plugin.cc]: An inline function has been created.

**Comment by guiccbr** [comment]
@dakerfp and @rchiossi, do you have anything else to add about the fixes I've submitted?

**Pull request merged**